File: android/src/main/java/com/icure/kerberus/expo/ExpoKerberusModule.kt
package com.icure.kerberus.expo

import com.icure.keberus.Challenge
import com.icure.keberus.Solution
import com.icure.keberus.resolveChallenge
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

class ExpoKerberusModule : Module() {

  companion object {
    const val ON_PROGRESS_EVENT_NAME =  "onProgress"
  }

  override fun definition() = ModuleDefinition {
    Name("ExpoKerberus")

    Events(ON_PROGRESS_EVENT_NAME)

    AsyncFunction("resolveChallenge") Coroutine { challenge: ExpoChallenge, serializedInput: String ->
      resolveChallenge(
        challenge.toChallenge(),
        serializedInput
      ){
        sendEvent(ON_PROGRESS_EVENT_NAME, mapOf("progress" to it))
      }.toExpo()
    }
  }
}

class ExpoChallenge(
  @Field val id: String,
  @Field val salts: List<String>,
  @Field val difficultyFactor: Int
): Record {
  internal fun toChallenge() = Challenge(
    id = id,
    salts = salts,
    difficultyFactor = difficultyFactor
  )
}

class ExpoSolution(
  @Field val id: String,
  @Field val nonces: List<String>,
): Record

internal fun Solution.toExpo() = ExpoSolution(
  id = id,
  nonces = nonces
)

File: ios/ExpoKerberusModule.swift
import ExpoModulesCore
import KerberusKotlin
import Foundation

public class ExpoKerberusModule: Module {
    
  // Each module class must implement the definition function. The definition consists of components
  // that describes the module's functionality and behavior.
  // See https://docs.expo.dev/modules/module-api for more details about available components.
  public func definition() -> ModuleDefinition {
    // Sets the name of the module that JavaScript code will use to refer to the module. Takes a string as an argument.
    // Can be inferred from module's class name, but it's recommended to set it explicitly for clarity.
    // The module will be accessible from `requireNativeModule('ExpoKerberus')` in JavaScript.
    Name("ExpoKerberus")
      
    Events("onProgress")

    // Defines a JavaScript function that always returns a Promise and whose native code
    // is by default dispatched on the different thread than the JavaScript runtime runs on.
      AsyncFunction("resolveChallenge") { (challenge: ExpoChallenge, serializedInput: String) -> ExpoSolution in
          try await KerberusKt.resolveChallenge(config: challenge.toChallenge(), serializedInput: serializedInput) { progress in
              self.sendEvent("onProgress", ["progress": progress.doubleValue])
          }.toExpoSolution()
    }
  }
}

struct ExpoChallenge: Record {
    @Field
    var id: String
    @Field
    var salts: [String]
    @Field
    var difficultyFactor: Int

    func toChallenge() -> Challenge {
        return Challenge(id: id, salts: salts, difficultyFactor: Int32(difficultyFactor))
    }
}

struct ExpoSolution: Record {
    @Field
    var id: String
    @Field
    var nonces: [String]
}

extension Solution {
    func toExpoSolution() -> ExpoSolution {
        return ExpoSolution(id: Field(wrappedValue: id), nonces: Field(wrappedValue: nonces))
    }
}
